File: mobile_client/lib/services/game/timer_service.dart
import 'package:flutter/foundation.dart';
import 'package:mobile_client/constants/enum/timer_type.dart';
import 'package:mobile_client/constants/socket_events/timer_event.dart';
import 'package:mobile_client/services/websocket_service.dart';
import 'package:mobile_client/utils/void_notifier.dart';

class TimerService {
  TimerService(this.webSocketService);

  ValueNotifier<int> time = ValueNotifier<int>(0);
  ValueNotifier<TimerType> timerStarted =
      ValueNotifier<TimerType>(TimerType.none);
  VoidNotifier timerEnded = VoidNotifier();

  ValueNotifier<bool> isTimerActivated = ValueNotifier(false);

  ValueNotifier<bool> isPanicMode = ValueNotifier(false);
  TimerType currentType = TimerType.none;

  final WebSocketService webSocketService;

  void setTimerListeners() {
    removeTimerListeners();
    _setEventListeners();
  }

  void removeTimerListeners() {
    webSocketService.removeAllListeners(TimerEvent.panicTimer.name);
    webSocketService.removeAllListeners(TimerEvent.timerTick.name);
    webSocketService.removeAllListeners(TimerEvent.timerEnded.name);
    webSocketService.removeAllListeners(TimerEvent.timerStarted.name);
    currentType = TimerType.none;
    isPanicMode.value = false;
    time.value = 0;
    removeResetValueNotifiers();
  }

  void removeResetValueNotifiers() {
    timerStarted.dispose();
    time.dispose();
    timerEnded.dispose();
    isTimerActivated.dispose();

    time = ValueNotifier<int>(0);
    timerStarted = ValueNotifier<TimerType>(TimerType.none);
    timerEnded = VoidNotifier();
    isTimerActivated = ValueNotifier(false);
  }

  void pauseCountdown() {
    isTimerActivated.value = false;
    webSocketService.send(TimerEvent.stopTimer.name);
  }

  void enterPanicMode() {
    isTimerActivated.value = true;
    webSocketService.send(TimerEvent.panicTimer.name);
  }

  void resumeCountdown() {
    webSocketService.send(TimerEvent.resumeTimer.name);
  }

  void _setEventListeners() {
    webSocketService.on<dynamic>(TimerEvent.timerStarted.name, (data) {
      currentType = TimerType.fromJson(data);
      isTimerActivated.value = true;
      timerStarted.value = currentType;
    });

    webSocketService.on<void>(TimerEvent.timerEnded.name, (_) {
      isTimerActivated.value = false;
      isPanicMode.value = false;
      timerEnded.notify();
    });

    webSocketService.on<dynamic>(TimerEvent.timerTick.name, (timeValue) {
      isTimerActivated.value = true;
      time.value = timeValue;
    });

    webSocketService.on<void>(TimerEvent.panicTimer.name, (_) {
      // TO DO ADD panic audio
      isPanicMode.value = true;
    });
  }
}

File: desktop-client/src/app/game/services/timer/timer.service.ts
import { Injectable } from '@angular/core';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { TimerType } from '@common/enum/timer-type';
import { TimerEvent } from '@common/socket-events/timer-event';
import { Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class TimerService {
    time: number = 0;
    isTimerActivated: boolean = false;
    isPanicMode: boolean = false;
    timerEnded = new Subject<void>();
    timerStarted = new Subject<TimerType | undefined>();
    currentType: TimerType | undefined = TimerType.GameStart;

    constructor(private readonly webSocketService: WebsocketService) {}

    setTimerListeners() {
        this.removeTimerListeners();
        this.setEventListeners();
    }

    removeTimerListeners() {
        this.webSocketService.removeAllListeners(TimerEvent.PANIC_TIMER);
        this.webSocketService.removeAllListeners(TimerEvent.TIMER_TICK);
        this.webSocketService.removeAllListeners(TimerEvent.TIMER_ENDED);
        this.webSocketService.removeAllListeners(TimerEvent.TIMER_STARTED);
    }

    pauseCountdown() {
        this.isTimerActivated = false;
        this.webSocketService.send(TimerEvent.STOP_TIMER);
    }

    enterPanicMode() {
        this.isTimerActivated = true;
        this.webSocketService.send(TimerEvent.PANIC_TIMER);
    }

    resumeCountdown() {
        this.webSocketService.send(TimerEvent.RESUME_TIMER);
    }

    resetTimer() {
        this.isTimerActivated = false;
        this.time = 0;
        this.isPanicMode = false;
        this.currentType = undefined;
        this.resetSubjects();
    }

    resetSubjects() {
        this.timerStarted.complete();
        this.timerEnded.complete();
        this.timerEnded = new Subject<void>();
        this.timerStarted = new Subject<TimerType | undefined>();
    }

    private setEventListeners() {
        this.webSocketService.on<TimerType>(TimerEvent.TIMER_STARTED, (data?: TimerType) => {
            this.isTimerActivated = true;
            this.isPanicMode = false;
            this.currentType = data;
            this.timerStarted.next(data);
        });
        this.webSocketService.on<void>(TimerEvent.TIMER_ENDED, () => {
            this.isTimerActivated = false;
            this.isPanicMode = false;
            this.timerEnded.next();
        });
        this.webSocketService.on<number>(TimerEvent.TIMER_TICK, (time?: number) => {
            this.isTimerActivated = true;
            if (time !== undefined) this.time = time;
        });
        this.webSocketService.on<void>(TimerEvent.PANIC_TIMER, () => {
            this.isPanicMode = true;
        });
    }
}
